File: BaekJoon/Silver/2108/Python/2108.py
import sys

n = int(sys.stdin.readline())
arr1 = []
memo = {}
for i in range(0, n):
    num = int(sys.stdin.readline())
    if num not in memo:
        memo[num] = 1
    else:
        memo[num] += 1
    arr1.append(num)
arr1.sort()

sum = 0
for val in arr1:
    sum += val

average = round(sum / n)
median = arr1[(0+len(arr1)-1)//2]
range = arr1[-1] - arr1[0]
mode_keys = []
mode_max = 0
for key in memo:
    if memo[key] > mode_max:
        mode_keys = []
        mode_max = memo[key]
        mode_keys.append(key)
    elif memo[key] == mode_max:
        mode_keys.append(key)
    else:
        pass
mode_keys.sort()

mode = mode_keys[0] if len(mode_keys) == 1 else mode_keys[1]
print(average, median, mode, range, sep='\n')

File: BaekJoon/Silver/2108/C++/2108.cpp
// no.2108: 통계학 (S3)

#include <cstdio>
#include <algorithm>
#include <unordered_map>
#include <cmath>
#include <vector>
using namespace std;

int main() {
    int n;
    scanf("%d", &n);
    vector<int> arr;
    unordered_map<int, int> cnt;
    for(int i=0; i<n; i++) {
        int num;
        scanf("%d", &num);
        arr.push_back(num);
        if(cnt.count(num) == 0)
            cnt[num] = 1;
        else
            cnt[num]++;
    }
    sort(arr.begin(), arr.end());
    int sum = 0;
    for(int num : arr) {
        sum += num;
    }
    int avg = (int)floor((double)sum/n+0.5);
    int median = arr[n/2];
    int max = 0;
    for(pair<int, int> tmp : cnt) {
        if(tmp.second > max) {
            max = tmp.second;
        }
    }
    vector<int> modes;
    for(pair<int, int> tmp : cnt) {
        if(tmp.second == max)
            modes.push_back(tmp.first);
    }
    int mode;
    if(modes.size() == 1) {
        mode = modes[0];
    } else {
        sort(modes.begin(), modes.end());
        mode = modes[1];
    }
    int range = arr[n-1]-arr[0];
    printf("%d\n", avg);
    printf("%d\n", median);
    printf("%d\n", mode);
    printf("%d\n", range);
    return 0;
}